### PR TITLE
[DRTV extractor] Fix drtv.py

### LINKED ISSUE #8298: [DRTV] Unable to download JSON metadata: HTTP Error 404 (dr.dk)
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I'm running yt-dlp version **2023.09.24** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Denmark

### Provide a description that is worded well enough to be understood

Downloading from "https://www.dr.dk/drtv/tv-guide" (Supported site "DRTV") was no problem 3 days ago. Since then, attempts to download previous broadcasts have resulted in the following error: 

_Unable to download JSON metadata: HTTP Error 404: Not Found (caused by <HTTPError 404: Not Found>)_

A test can be done with this url:
_yt-dlp.exe "https://www.dr.dk/drtv/se/alene-i-vildmarken_-opfindsomme-shelters_169521"_

Downloading something from "https://www.dr.dk/drtv/tv-guide" 5 minutes after a direct transmission ends is no problem. I think that the transmission then is transcoded into different formats, and these transcoded formats can no longer be downloaded because of a missing JSON.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
yt-dlp.exe -vU "https://www.dr.dk/drtv/se/alene-i-vildmarken_-opfindsomme-shelters_169521"
[debug] Command-line config: ['-vU', 'https://www.dr.dk/drtv/se/alene-i-vildmarken_-opfindsomme-shelters_169521']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.10.04.180415 [b095fd3fa] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.22621-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 2022-03-07-git-e645a1ddb9-essentials_build-www.gyan.dev (setts), ffprobe 2022-03-07-git-e645a1ddb9-essentials_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2023.07.22, mutagen-1.47.0, sqlite3-3.35.5, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1886 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.10.04.180415, Current version: nightly@2023.10.04.180415
Current Build Hash: 00fc0b80ab8bf1f1d3acc7002a6f078bb0574f63e6a6e2dcf18d82e736c9f482
yt-dlp is up to date (nightly@2023.10.04.180415)
[drtv] Extracting URL: https://www.dr.dk/drtv/se/alene-i-vildmarken_-opfindsomme-shelters_169521
[drtv] alene-i-vildmarken_-opfindsomme-shelters_169521: Downloading webpage
[drtv] 00922000810: Downloading video JSON
ERROR: [drtv] alene-i-vildmarken_-opfindsomme-shelters_169521: Unable to download JSON metadata: HTTP Error 404: Not Found (caused by <HTTPError 404: Not Found>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 715, in extract
  File "yt_dlp\extractor\drtv.py", line 203, in _real_extract
  File "yt_dlp\extractor\common.py", line 1069, in download_content
  File "yt_dlp\extractor\common.py", line 1033, in download_handle
  File "yt_dlp\extractor\common.py", line 903, in _download_webpage_handle
  File "yt_dlp\extractor\common.py", line 860, in _request_webpage

  File "yt_dlp\networking\_urllib.py", line 410, in _send
  File "urllib\request.py", line 531, in open
  File "urllib\request.py", line 640, in http_response
  File "urllib\request.py", line 569, in error
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 649, in http_error_default
urllib.error.HTTPError: HTTP Error 404: Not Found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 4051, in urlopen
  File "yt_dlp\networking\common.py", line 114, in send
  File "yt_dlp\networking\_helper.py", line 204, in wrapper
  File "yt_dlp\networking\common.py", line 325, in send
  File "yt_dlp\networking\_urllib.py", line 415, in _send
yt_dlp.networking.exceptions.HTTPError: HTTP Error 404: Not Found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "yt_dlp\extractor\common.py", line 847, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 4070, in urlopen
yt_dlp.networking.exceptions._CompatHTTPError: HTTP Error 404: Not Found
```

### PR DESCRIPTION
**Pull request by almx**
### Description of your *pull request* and other information

Changed code to be able to download from dr.dk's new format.

Fixes #8298

Code taken from https://gist.github.com/lyngklip/3be04b1cf380536f76713cd09d4ac61a

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🔄🗑️🔢

Update DRTV extractor to use new API and JSON format. Simplify and clean up the code.

> _`DRTV` videos_
> _New endpoint, JSON data_
> _Old code falls like leaves_

### Walkthrough
* Rewrite extraction logic for DRTV videos using a different API endpoint and JSON data structure ([link](https://github.com/yt-dlp/yt-dlp/pull/8484/files?diff=unified&w=0#diff-0363d92c142b50b88cfc0b912879c30a0dfb720feb57f35ae1d2e2073e9d226bL164-R212))
* Add imports of `json`, `uuid`, and `compat_etree_fromstring` for the new extraction logic ([link](https://github.com/yt-dlp/yt-dlp/pull/8484/files?diff=unified&w=0#diff-0363d92c142b50b88cfc0b912879c30a0dfb720feb57f35ae1d2e2073e9d226bL3-R9))
* Remove unused constant and helper functions that were based on the deprecated and encrypted API endpoint ([link](https://github.com/yt-dlp/yt-dlp/pull/8484/files?diff=unified&w=0#diff-0363d92c142b50b88cfc0b912879c30a0dfb720feb57f35ae1d2e2073e9d226bL22))
* Move `DRTVLiveIE` class to the end of the file to follow alphabetical order of extractors ([link](https://github.com/yt-dlp/yt-dlp/pull/8484/files?diff=unified&w=0#diff-0363d92c142b50b88cfc0b912879c30a0dfb720feb57f35ae1d2e2073e9d226bL403))

</details>

### CODE REVIEW

File: yt_dlp/extractor/drtv.py [hunk 1]
```diff
@@ -161,191 +162,54 @@ class DRTVIE(InfoExtractor):
...
-                    })
-
-        if not formats and restricted_to_denmark:
-            self.raise_geo_restricted(
-                'Unfortunately, DR is not allowed to show this program outside Denmark.',
-                countries=self._GEO_COUNTRIES)
-
+        for fmt in data:
+            formatId = fmt['format'] or 'na'
+            accessService = fmt['accessService']
```
Comment (seproDev) [inline yt_dlp/extractor/drtv.py:180]: Use snake_case and make this non-fatal:
```suggestion
            format_id = fmt.get('format', 'na')
            access_service = fmt.get('accessService')
```

File: yt_dlp/extractor/drtv.py [hunk 2]
```diff
@@ -161,191 +162,54 @@ class DRTVIE(InfoExtractor):
...
+            if accessService in ('SpokenSubtitles', 'SignLanguage', 'VisuallyInterpreted'):
+                preference = -1
+                formatId += '-%s' % accessService
+            elif accessService == 'StandardVideo':
+                preference = 1
+            fmts, subs = self._extract_m3u8_formats_and_subtitles(fmt['url'], videoId, 'mp4', entry_protocol='m3u8_native', preference=preference, m3u8_id=formatId, fatal=False)
+            formats.extend(fmts)
+            self._merge_subtitles(subs, target=subtitles)
+            LANGS = {
+                'Danish': 'da',
```
Comment (seproDev) [inline yt_dlp/extractor/drtv.py:191]: I only checked a single video, but for `169521` the subtitle language is set to `DanishLanguageSubtitles`. Do both need to be accepted, and are there other possible values?

**Review by seproDev — CHANGES_REQUESTED** [review]
Partial review for now.
Make sure you actually run flake8 and the unit tests. Given your changes and the site seemingly changing URL structure, the unit tests will need to be updated.
Also take another look at the [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md) for things like the long line policy.

File: yt_dlp/extractor/drtv.py (hunk 1 above)
Comment (almx) [reply yt_dlp/extractor/drtv.py:180]: Will fix

**Comment by almx** [comment]
Thanks for the feedback, I am in the process of implementing the requested changes.

File: yt_dlp/extractor/drtv.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 import binascii
 import hashlib
+import json
 import re
+import uuid
 
 from .common import InfoExtractor
 from ..aes import aes_cbc_decrypt_bytes, unpad_pkcs7
-from ..compat import compat_urllib_parse_unquote
+from ..compat import compat_etree_fromstring, compat_urllib_parse_unquote
```
[the remark this answers is not in the document]
Comment (almx) [reply yt_dlp/extractor/drtv.py:9]: fixed

File: yt_dlp/extractor/drtv.py [hunk 4]
```diff
@@ -161,191 +162,54 @@ class DRTVIE(InfoExtractor):
...
-        thumbnail = None
-        duration = None
-
-        restricted_to_denmark = False
-
+        json_data = self._search_json(r'window\.__data\s*=', webpage, 'data', raw_video_id)
+        item = traverse_obj(json_data, ('cache', 'page', Ellipsis, (None, ('entries', 0)), 'item'), get_all=False)
+        itemId = item.get('id')
+        videoId = item['customId'].split(':')[-1]
+        deviceId = uuid.uuid1()
```
[the remark this answers is not in the document]
Comment (almx) [reply yt_dlp/extractor/drtv.py:173]: fixed

File: yt_dlp/extractor/drtv.py [hunk 5]
```diff
@@ -161,191 +162,54 @@ class DRTVIE(InfoExtractor):
...
-            data.get('PrimaryBroadcastStartTime') or data.get('SortDateTime'))
-
-        thumbnail = None
-        duration = None
-
-        restricted_to_denmark = False
-
+        json_data = self._search_json(r'window\.__data\s*=', webpage, 'data', raw_video_id)
+        item = traverse_obj(json_data, ('cache', 'page', Ellipsis, (None, ('entries', 0)), 'item'), get_all=False)
+        itemId = item.get('id')
```
[the remark this answers is not in the document]
Comment (almx) [reply yt_dlp/extractor/drtv.py:171]: fixed

File: yt_dlp/extractor/drtv.py [hunk 6]
```diff
@@ -161,191 +162,54 @@ class DRTVIE(InfoExtractor):
 
     def _real_extract(self, url):
         raw_video_id, is_radio_url = self._match_valid_url(url).group('id', 'radio')
```
[the remark this answers is not in the document]
Comment (almx) [reply yt_dlp/extractor/drtv.py:164]: I have removed is_radio_url as I am, at a glance, unable to make radio (sound/podcast) extraction work.

File: yt_dlp/extractor/drtv.py [hunk 7]
```diff
@@ -161,191 +162,54 @@ class DRTVIE(InfoExtractor):
...
-        timestamp = unified_timestamp(
-            data.get('PrimaryBroadcastStartTime') or data.get('SortDateTime'))
-
-        thumbnail = None
-        duration = None
-
-        restricted_to_denmark = False
-
+        json_data = self._search_json(r'window\.__data\s*=', webpage, 'data', raw_video_id)
+        item = traverse_obj(json_data, ('cache', 'page', Ellipsis, (None, ('entries', 0)), 'item'), get_all=False)
```
[the remark this answers is not in the document]
Comment (almx) [reply yt_dlp/extractor/drtv.py:170]: I tried your code with the ..., but it didn't work. I am not well-versed enough in Python to change it much.

File: yt_dlp/extractor/drtv.py [hunk 8]
```diff
@@ -19,7 +21,6 @@
 
 SERIES_API = 'https://production-cdn.dr-massive.com/api/page?device=web_browser&item_detail_expand=all&lang=da&max_list_prefetch=3&path=%s'
 
-
```
[the remark this answers is not in the document]
Comment (almx) [reply yt_dlp/extractor/drtv.py:22]: Done

File: yt_dlp/extractor/drtv.py [hunk 9]
```diff
@@ -161,191 +162,54 @@ class DRTVIE(InfoExtractor):
...
-        thumbnail = None
-        duration = None
-
-        restricted_to_denmark = False
-
+        json_data = self._search_json(r'window\.__data\s*=', webpage, 'data', raw_video_id)
+        item = traverse_obj(json_data, ('cache', 'page', Ellipsis, (None, ('entries', 0)), 'item'), get_all=False)
+        itemId = item.get('id')
+        videoId = item['customId'].split(':')[-1]
```
[the remark this answers is not in the document]
Comment (almx) [reply yt_dlp/extractor/drtv.py:172]: Perhaps, but I'm thinking it's more future proof to extract the ID from the JSON?

File: yt_dlp/extractor/drtv.py (hunk 2 above)
Comment (almx) [reply yt_dlp/extractor/drtv.py:191]: Fixed.

The subtitle language can also have other values such as ForeignLanguageSubtitles and CombinedLanguageSubtitles in the case non-Danish is spoken in the video. They will always be ultimately Danish language (code 'da'), but when using --all-subs each sub needs to have a unique language code, so I can't really think of a solution here other than let them have the raw name (for example FunnyShow.CombinedLanguageSubtitles .vtt)

File: yt_dlp/extractor/drtv.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bashonly) [reply yt_dlp/extractor/drtv.py:151]: the `None` branch is significant, so all that we need to change is `Ellipsis` to `...`:
```suggestion
        item = traverse_obj(json_data, ('cache', 'page', ..., (None, ('entries', 0)), 'item'), get_all=False)
```

[the remark this answers is not in the document]
Comment (seproDev) [reply yt_dlp/extractor/drtv.py:151]: Oh, I hadn't noticed that some of the pages have differently structured hydration data.
Thanks for pointing that out.

**Comment by seproDev** [comment]
Okay, so I mostly fixed up `DRTVIE`. There is still a fair bit of stuff to be done:
- `DRTVIE` will need to be more extensively tested, especially regarding multiple video/subtitle tracks. Two different danish subtitle tracks are currently not possible. More test cases should be added.
- `DRTVSeasonIE` and `DRTVSeriesIE` will need to be fixed. From a quick glance, it seems like the same auth token is required, so moving `_real_initialize` to a base class probably makes sense.
- Nyheder and radio/lyd will need its own extractor as the APIs are too different now.

For this PR to get merged, I think at least the first two need to be addressed. I might work on this more when I have the time.

File: yt_dlp/extractor/drtv.py [hunk 10]
```diff
@@ -200,19 +200,24 @@ def _real_extract(self, url):
             format_id = fmt.get('format', 'na')
             access_service = fmt.get('accessService')
             preference = None
+            subtitle_sufix = ''
```
Comment (barsnick) [inline yt_dlp/extractor/drtv.py:203]: sufix -> suffix

File: yt_dlp/extractor/drtv.py [hunk 11]
```diff
@@ -24,7 +19,7 @@ class DRTVIE(InfoExtractor):
     _VALID_URL = r'''(?x)
                     https?://
                         (?:
-                            (?:www\.)?dr\.dk/(?:tv/se|nyheder|(?P<radio>radio|lyd)(?:/ondemand)?)/(?:[^/]+/)*|
+                            (?:www\.)?dr\.dk/tv/se(?:/ondemand)?/(?:[^/]+/)*|
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:22]: exclude query and fragment from the match
```suggestion
                            (?:www\.)?dr\.dk/tv/se(?:/ondemand)?/(?:[^/?#]+/)*|
```

File: yt_dlp/extractor/drtv.py [hunk 12]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
-                video_id = compat_urllib_parse_unquote(video_id)
+    def _real_initialize(self):
+        if self._TOKEN:
+            return
+        token_response = self._download_json(
+            'https://production.dr-massive.com/api/authorization/anonymous-sso', None,
+            note='Downloading anonymous token', headers={
+                'content-type': 'application/json',
+            },
+            query={
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:144]:
```suggestion
            }, query={
```

File: yt_dlp/extractor/drtv.py [hunk 13]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
+                'device': 'web_browser',
+                'ff': 'idp,ldp,rpt',
+                'lang': 'da',
+                'supportFallbackToken': 'true',
+            },
+            data=json.dumps({
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:150]:
```suggestion
            }, data=json.dumps({
```

File: yt_dlp/extractor/drtv.py [hunk 14]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
+                'deviceId': str(uuid.uuid4()),
+                'scopes': [
+                    'Catalog',
+                ],
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:154]:
```suggestion
                'scopes': ['Catalog'],
```

File: yt_dlp/extractor/drtv.py [hunk 15]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
+                'optout': True,
+            }).encode())
+
+        self._TOKEN = traverse_obj(
+            token_response, (lambda _, x: x['type'] == 'UserAccount', 'value', {str}), get_all=False)
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:160]: this could fail silently; shouldn't we raise `if not self._TOKEN`?

File: yt_dlp/extractor/drtv.py [hunk 16]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
 
-        _PROGRAMCARD_BASE = 'https://www.dr.dk/mu-online/api/1.4/programcard'
-        query = {'expanded': 'true'}
+    def _real_extract(self, url):
+        url_slug = self._match_id(url)
+        webpage = self._download_webpage(url, url_slug)
 
-        if video_id:
-            programcard_url = '%s/%s' % (_PROGRAMCARD_BASE, video_id)
+        json_data = self._search_json(r'window\.__data\s*=\s*', webpage, 'data', url_slug, fatal=False) or {}
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:165]: `_search_json` appends `\s*` to the start pattern already
```suggestion
        json_data = self._search_json(
            r'window\.__data\s*=', webpage, 'data', url_slug, fatal=False) or {}
```

File: yt_dlp/extractor/drtv.py [hunk 17]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
+        item = traverse_obj(json_data, ('cache', 'page', ..., (None, ('entries', 0)), 'item'), get_all=False)
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:166]: expect the type
```suggestion
        item = traverse_obj(
            json_data, ('cache', 'page', ..., (None, ('entries', 0)), 'item', {dict}), get_all=False)
```

File: yt_dlp/extractor/drtv.py [hunk 18]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
+                    'device': 'web_browser',
+                    'expand': 'all',
+                    'ff': 'idp,ldp,rpt',
+                    'geoLocation': 'dk',
+                    'isDeviceAbroad': 'false',
+                    'lang': 'da',
+                    'segments': 'drtv,optedout',
+                    'sub': 'Anonymous',
+                })
+        video_id = try_call(lambda: item['customId'].rsplit(':', 1)[-1]) or item_id
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:183]: add newline for readability
```suggestion
                })

        video_id = try_call(lambda: item['customId'].rsplit(':', 1)[-1]) or item_id
```

File: yt_dlp/extractor/drtv.py [hunk 19]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
+                'delivery': 'stream',
+                'device': 'web_browser',
+                'ff': 'idp,ldp,rpt',
+                'lang': 'da',
+                'resolution': 'HD-1080',
+                'sub': 'Anonymous',
+            },
+            headers={
+                'authorization': f'Bearer {self._TOKEN}',
+            })
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:196]:
```suggestion
            }, headers={'authorization': f'Bearer {self._TOKEN}'})
```

File: yt_dlp/extractor/drtv.py [hunk 20]
```diff
@@ -123,226 +123,125 @@ class DRTVIE(InfoExtractor):
...
-                    subtitles.setdefault(LANGS.get(lang, lang), []).append({
-                        'url': sub_uri,
-                        'ext': mimetype2ext(subs.get('MimeType')) or 'vtt'
-                    })
-
-        if not formats and restricted_to_denmark:
-            self.raise_geo_restricted(
-                'Unfortunately, DR is not allowed to show this program outside Denmark.',
-                countries=self._GEO_COUNTRIES)
+        for fmt in traverse_obj(stream_data, (lambda _, x: x['url'], {dict})):
```
Comment (bashonly) [inline yt_dlp/extractor/drtv.py:200]: this is enough to determine it's a dict
```suggestion
        for fmt in traverse_obj(stream_data, (lambda _, x: x['url'])):
```
&nbsp;
More importantly, I'm not a fan of using `fmt` as the variable name here... can be easily confused with fmts/formats.
`stream` would be better IMO

**Review by bashonly — CHANGES_REQUESTED** [review]
Some questions:

Are we not able to detect/raise for geo restriction any more? Seems important

Is it assumed that the radio/lyd links will need their own extractor and that's why support has been removed from `DRTVIE`?

**Comment by seproDev** [comment]
> Are we not able to detect/raise for geo restriction any more? Seems important

For TV shows, there is a flag in the api response to see if the show is geo restricted, so I added a check for that.
The browser just blindly tries the m3u8 URLs and if unsuccessful displays the geo restricted error, even if the m3u8 file 404ed for example.

> Is it assumed that the radio/lyd links will need their own extractor and that's why support has been removed from DRTVIE?

Yes, the APIs used are too different now.

**Review by bashonly — APPROVED** [review]

**Pull request merged**